### PR TITLE
Added support for multi-qubit gates in qcircuit

### PR DESCRIPTION
**Pull request by bryano**

### CODE REVIEW

File: cirq/contrib/qcircuit/qcircuit_test.py [hunk 1]
```diff
@@ -34,13 +34,66 @@ def __str__(self):
...
+    f = _TextToQCircuitDiagrammable(g)
+    qubit_map = {q: i for q, i in zip(qubits, (4, 3))}
+    actual_info = f.qcircuit_diagram_info(
+            cirq.TextDiagramInfoArgs(known_qubits=qubits,
+                                     known_qubit_count=None,
+                                     use_unicode_characters=True,
+                                     precision=3,
+                                     qubit_map=qubit_map))
+    expected_info = cirq.TextDiagramInfo(
+            ('\ghost{\\text{SWAP}}', '\multigate{1}{\\text{SWAP}}'),
```
[the remark this answers is not in the document]
Comment (bryano) [reply cirq/contrib/qcircuit/qcircuit_test.py]: Yea, this was just to test the functionality; any InterchangeableQubitGate would do.

File: cirq/contrib/qcircuit/qcircuit_diagrammable.py [hunk 2]
```diff
@@ -12,17 +12,16 @@
...
 from cirq import Extensions, ops
 from cirq import abc
+from cirq.ops import gate_features
 
 
 class QCircuitDiagrammable(metaclass=abc.ABCMeta):
     @abc.abstractmethod
     def qcircuit_diagram_info(self, args: ops.TextDiagramInfoArgs
-                              ) -> Tuple[str, ...]:
+                              ) -> ops.TextDiagramInfo:
```
[the remark this answers is not in the document]
Comment (bryano) [reply cirq/contrib/qcircuit/qcircuit_diagrammable.py]: Master still has qcircuit_diagram_info -> Tuple.

[the remark this answers is not in the document]
Comment (bryano) [reply cirq/contrib/qcircuit/qcircuit_diagrammable.py]: Which, BTW, is fine if the only thing that changes between text diagrams and qcircuit diagrams is the wire symbols. But that's not the case here.

File: cirq/contrib/qcircuit/qcircuit_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bryano) [reply cirq/contrib/qcircuit/qcircuit_test.py]: In general, what's the policy on what information needs to be in diagrams? It seems that as of now, the circuit is always exactly deducible from the diagram, but that's not always the most desirable. For example, if you have a two-qubit gate U that's not interchangeable, maybe for some purposes you just want a box with a U in it, even if technically that hides information about the ordering of the qubits that U acts on. Maybe a 'verbose' flag in TextDiagramInfoArgs that indicates whether to be exact or pretty?

**Comment by bryano** [comment]
Mypy isn't smart enough to see that the relevant attributes are not None at the point in the code that they're referenced. Is there an annotation to help it out?

File: cirq/contrib/qcircuit/qcircuit_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/contrib/qcircuit/qcircuit_test.py]: I would instead call that flag "allow_ambiguous".

File: cirq/contrib/qcircuit/qcircuit_diagrammable.py [hunk 3]
```diff
@@ -46,36 +47,63 @@ def __init__(self, *symbols: str) -> None:
...
     def __init__(self, sub: ops.TextDiagrammable) -> None:
         self.sub = sub
 
     def qcircuit_diagram_info(self, args: ops.TextDiagramInfoArgs
-                              ) -> Tuple[str, ...]:
+                              ) -> ops.TextDiagramInfo:
         info = self.sub.text_diagram_info(args)
-
+        multigate_parameters = _get_multigate_parameters(self.sub, args)
+        if multigate_parameters is not None:
```
Comment (Strilanc) [inline cirq/contrib/qcircuit/qcircuit_diagrammable.py]: Put this `if` block at the end of `_get_multigate_parameters` and have that method return `Optional[ops.TextDiagramInfo]`. That will likely fix the type error.

**Comment by Strilanc** [comment]
You can use `cast` from `typing` to assert that something is a type. E.g. `cast(float, x)`.

**Comment by bryano** [comment]
Casting doesn't seem to work, even if I define an intermediate variable. Not sure why.

Are the assertions not okay?

**Comment by Strilanc** [comment]
The assertions are fine.

**Pull request merged**